logins-sql/src/db.rs

```diff
@@ -0,0 +1,789 @@
+                }
+                seen_ids.insert(incoming.0.id.clone());
+                sync_data.push(SyncLoginData::from_payload(incoming.0.clone(), incoming.1)?);
+            }
+        }
+
+        util::each_chunk_mapped(&records, self.max_var_count, |r| &r.0.id as &ToSql, |chunk, offset| {
+            // pairs the bound parameter for the guid with an integer index.
+            let values_with_idx = util::repeat_display(chunk.len(), ",", |i, f| write!(f, "({},?)", i + offset));
+            let query = format!("
+                WITH to_fetch(guid_idx, fetch_guid) AS (VALUES {vals})
+                SELECT
+                    {common_cols},
+                    is_overridden,
+                    server_modified,
+                    NULL as local_modified,
+                    NULL as is_deleted,
+                    NULL as sync_status,
+                    1 as is_mirror,
+                    to_fetch.guid_idx as guid_idx
+                FROM {mirror_table}
+                JOIN to_fetch
+                  ON {mirror_table}.guid = to_fetch.fetch_guid
+
+                UNION ALL
+
+                SELECT
+                    {common_cols},
+                    NULL as is_overridden,
+                    NULL as server_modified,
+                    local_modified,
+                    is_deleted,
+                    sync_status,
+                    0 as is_mirror,
+                    to_fetch.guid_idx as guid_idx
+                FROM {local_table}
+                JOIN to_fetch
+                  ON {local_table}.guid = to_fetch.fetch_guid",
+                // give each VALUES item 2 entries, an index and the parameter.
+                vals = values_with_idx,
+                local_table = schema::LOCAL_TABLE_NAME,
+                mirror_table = schema::MIRROR_TABLE_NAME,
+                common_cols = schema::COMMON_COLS
+            );
+
+            let mut stmt = self.db.prepare(&query)?;
+
+            let rows = stmt.query_and_then(chunk, |row| {
+                let guid_idx_i = row.get::<_, i64>("guid_idx");
+                // Hitting this means our math is wrong...
+                assert!(guid_idx_i >= 0);
+
+                let guid_idx = guid_idx_i as usize;
+                let is_mirror: bool = row.get("is_mirror");
+                if is_mirror {
+                    sync_data[guid_idx].set_mirror(MirrorLogin::from_row(row)?)?;
+                } else {
+                    sync_data[guid_idx].set_local(LocalLogin::from_row(row)?)?;
+                }
+                Ok(())
+            })?;
+            // `rows` is an Iterator<Item = Result<()>>, so we need to collect to handle the errors.
+            rows.collect::<Result<_>>()?;
+            Ok(())
+        })?;
+        Ok(sync_data)
+    }
+
+    // It would be nice if this were a batch-ish api (e.g. takes a slice of records and finds dupes
+    // for each one if they exist)... I can't think of how to write that query, though.
```

Comment:
You could stuff 🥙  the candidates into a temp table, then join on it to find dupes...but, IIUC, `reconcile` operates on one record at a time? Or, everything in "local" is fair game for duping? If so, do you need to loop over the records or the temp table at all?

Reply:
> but, IIUC, reconcile operates on one record at a time

I think it would be a new parameter in the UpdatePlan.

That said, I'm not really that concerned about this and it's probably clearer as is.